### ISSUE TITLE
Optimization of Array.concat()

### ISSUE THREAD

**Issue by dan1wang** [google/closure-compiler#3211]
It seems Closure Compiler (ADVANCED_OPTIMIZATIONS) is not very good at Array concatenate.

If I have the following (formatted for readability):
```
const template4 = [].concat(
        offsetSegment(array1, r1x, r1y),
        offsetSegment(array2, r4x, r4y),
        offsetSegment(array3, 0, obj.a),
        array4);
```

I would get
`  f=[].concat(offsetSegment(f,g,d),offsetSegment(m,g,n),offsetSegment(h,0,-a.b),k);`
instead of 
`  f=offsetSegment(f,g,d).concat(,offsetSegment(m,g,n),offsetSegment(h,0,-a.b),k);`

Similarly, if I have 
```
const template3 =
       array1
       .concat(offsetSegment(array2, r2x, r2y))
       .concat(offsetSegment(array2, r4x, r4y))
       .concat(offsetSegment(array3, 0, obj.a));
```

I would get something like
`  b=t.concat(offsetSegment(p,e,f)).concat(offsetSegment(m,n,o)).concat(offsetSegment(r,s,a.a,))`

In my externs file, I declared the various array functions with the following
```
/**
 * @nosideeffects
 * @param {Array} seg
 * @return {Array.<Array.<number>>}
 */
```

**Comment by tjgq** [comment]
Optimizing `[].concat(f(), x)` into `f().concat(x)` is only safe if `f` is guaranteed to return a fresh array every time it's called. `@nosideffects` is not enough to assert that: a side effect-free function may still, for example, return a reference to a non-local variable.

**Comment by dan1wang** [comment]
shouldn't x1 and x2 be the same regardless of whether f1() returns a fresh array?
```
x1 =[].concat(f1(), f2(), f3()...)
x2= f1().concat(f2(), f3(), ...)
```
If the code has a bug, wouldn't optimization just preserve the bug and not introduce a new one?

**Comment by brad4d** [comment]
@dan1wang you're right.
When I chatted with @tjgq about this I forgot that `concat` always creates a new array.

So, I suppose it should be safe to save 3 bytes by dropping the initial `[].` in such cases.
I feel certain that I'm forgetting some other concern that leads to us keeping it, though.

Feel free to take a look at the code if you're interested in tweaking this.
I don't think the closure-compiler team is likely to make it a priority, though.

**Comment by tjgq** [comment]
Thanks for the correction, Bradford. I too was under the mistaken impression that `concat` modified the array in place.

Created Google internal issue b/123525196 for tracking.

**Comment by ChadKillingsworth** [comment]
This would be a good first contribution. It's possible to do as a peephole optimization.

**Comment by qqingvarqq** [comment]
I wanted to attempt this one.
Also, I noticed that this example can be optimized even more.
Code like this: 
1.
```javascript
const template3 =
       array1
       .concat(offsetSegment(array2, r2x, r2y))
       .concat(offsetSegment(array2, r4x, r4y))
       .concat(offsetSegment(array3, 0, obj.a));
```
can be optimized to this
2.
```javascript
const template3 = array1.concat(
             offsetSegment(array2, r2x, r2y),
             offsetSegment(array2, r4x, r4y),
             offsetSegment(array3, 0, obj.a)
);
```
because Array.concat side effect free, always create a new copy of an array and there no access to intermediate arrays. So the results of 1 and 2 are always identical.

**Comment by brad4d** [comment]
Actually, I think I remember now why we don't end up removing the `[].` of `[].concat(someArray, ....)`.

Very few of the peephole optimizations use type information.
`PeepholeFoldConstants.java` and `PeepholeReplaceKnownMethods.java` do if you have set the compiler option to use types for optimizations, but otherwise they don't.

I think this optimization of `concat` would go into `PeepholeReplaceKnownMethods.java` and would need to be guarded by its `useTypes` flag field.

**Comment by qqingvarqq** [comment]
@tjgq It can be closed via #3386
